### PR TITLE
Limit time spent in auction creation

### PR DESCRIPTION
**Pull request by vkgnosis**
This makes use of the streaming price estimator interface changes by streaming as many as prices as possible within a fixed time limit that we reserve for this. Orders for which we did not get prices in time are excluded. Since price estimation is cached it is likely that in the next round of auction creation more orders can be included.

### Test Plan

CI, new test

### CODE REVIEW

File: crates/orderbook/Cargo.toml [hunk 1]
```diff
@@ -52,3 +52,4 @@ web3 = { version = "0.18", default-features = false }
 secp256k1 = "0.21"
 mockall = "0.11"
 testlib = { path = "../testlib" }
+tokio = { version = "1.15", features = ["test-util"] }
```
Comment (nlordell) [inline crates/orderbook/Cargo.toml:55]: Aren't we on `tokio v1.17` now?

File: crates/orderbook/src/solvable_orders.rs [hunk 2]
```diff
@@ -327,41 +334,64 @@ impl Maintaining for SolvableOrdersCache {
...
+            let price = match to_normalized_price(price) {
+                Some(price) => price,
+                None => continue,
+            };
+            prices.insert(*token, price);
+        }
+    };
+    match tokio::time::timeout_at(deadline, collect_prices).await {
+        Ok(()) => (),
+        Err(_) => tracing::warn!(
```
Comment (nlordell) [inline crates/orderbook/src/solvable_orders.rs:366]: Can we maybe add some metrics to this to?
- How many orders are missing?
- How often the timeout happens

Can be in a separate PR as well.

File: crates/orderbook/src/solvable_orders.rs [hunk 3]
```diff
@@ -327,41 +334,64 @@ impl Maintaining for SolvableOrdersCache {
...
+            "auction native price collection took too long, got {} out of {}",
+            prices.len(),
+            traded_tokens.len()
+        ),
+    }
 
+    // Filter both orders and prices so that we only return orders that have prices and prices that
+    // have orders.
```
Comment (nlordell) [inline crates/orderbook/src/solvable_orders.rs:374]: Ah! This is important because we may have gotten the price for one of the tokens of an order, but not the other, and the token we have the price for only appears in that order. I.e. we are missing a token for an order, so we exclude the order; in excluding the order, we have a price with no equivalent order.

File: crates/orderbook/src/solvable_orders.rs [hunk 4]
```diff
@@ -327,41 +334,64 @@ impl Maintaining for SolvableOrdersCache {
...
-        .zip(traded_tokens)
-        .filter_map(|(price, token)| match price {
-            Ok(price) => Some((token, to_normalized_price(price)?)),
-            Err(err) => {
-                tracing::warn!(?token, ?err, "error estimating native token price");
-                None
-            }
-        })
-        .collect::<BTreeMap<_, _>>();
+    let mut prices = BTreeMap::new();
```
Comment (nlordell) [inline crates/orderbook/src/solvable_orders.rs:345]: Might make sense to have this be a `HashMap` for faster access. We return the `used_prices` which needs to be a `BTreeMap`.
```suggestion
    let mut prices = HashMap::new();
```

**Review by nlordell — APPROVED** [review]
Looks great!

Would like to see metrics for:
- How many times we timeout updating auctions
- How many orders we exclude because of missing prices.

This will help us keep an eye and understand just how often our batches are "incomplete".

File: crates/orderbook/Cargo.toml (hunk 1 above)
Comment (vkgnosis) [reply crates/orderbook/Cargo.toml:55]: I copied the version from another line. If tokio 1.17 is the newest version then we're probably already using it in the lock file since it is semver compatible.

**Review by MartinquaXD — APPROVED** [review]
I think the PR is fine as it is.

Just one idea: Did you consider updating the auction in a streaming manner after the timeout was reached. I think computing a big chunk of the auction like as it's done right now makes sense (to not return tiny auctions when somebody requests one right after we started updating) but theoretically the auction in the cache could get more orders as they come in.

This is probably overkill and too complicated so I don't really think we should go for it right now but maybe in the future this idea might be helpful.
